torchvision/transforms/_core.py

```diff
@@ -23,17 +24,40 @@ def query_sample(sample: Any, fn: Callable[[Any], Optional[T]]) -> Iterator[T]:
             yield result
 
 
+class TransformDispatch:
+    def __init__(self, transform):
+        self.transform = transform
+        self.transform.set_reset_auto(False)
+
+    def __call__(self, inputs):
+        if isinstance(inputs, collections.abc.Mapping):
+            inputs = {key: self.transform(value) for key, value in inputs.items()}
+        elif isinstance(inputs, collections.abc.Sequence):
+            inputs = tuple(self.transform(input) for input in inputs)
+        self.transform.wipeout_()
+        return inputs
```

Comment:
> This should be callable with a variable number of inputs, i.e. transform(image, bbox)

Is that necessary? I don't really see the point, a loop can do the job. Again, IMO it is weird to have a function that can take a dictionary _and_ any length of args. What happens if it gets both?

> The inputs might be nested containers, so IMO recursing through it is the easiest solution.

I think that there is a design choice that needs to be made here: is it important that anything can happen and that we take any possibility into account, or do we want to constrain the usage of our classes to a specific use case?
Are there datasets that have those complex composite structures, like `Tuple[Dict[str, Tensor], Tuple[Tensor, ...], Tensor]`? If that can happen i'm ok with recursive of course.

Reply:
> Are there datasets that have those complex composite structures

Yes. COCO for example looks like dict(image=Image(...), annotations=[dict(bbox=BoundingBox()), ...]). So you have a dict in a list in a dict. There might be even crazier constructs for datasets that we currently don't support. At this point there is no reason not to support parsing "all" nested containers. 

> > This should be callable with a variable number of inputs, i.e. transform(image, bbox)
> 
> Is that necessary? I don't really see the point, a loop can do the job. 

No, not necessary, but makes a good UX. "If the transform can parse single tensors and containers at the same time, why do I need to pass in two tensors in a container?"

Reply:
Ok agreed then, we can check if it is a tensor-like and pass that to transforms, otherwise call JointTransforms again.
But then in your example why do we need to unpack the list?

Wouldn't that work?

```python
class JointTransform:
    def __init__(self, transform):
        self.transform = transform
        self.transform.set_reset_auto(False)

    def __call__(self, inputs, _top=True):
        if isinstance(inputs, collections.abc.Mapping):
            inputs = {key: self(value, _top=False) for key, value in inputs.items()}
        elif isinstance(inputs, collections.abc.Sequence):
            inputs = tuple(self(input, _top=False) for input in inputs)
        else:
            inputs = self.transform(inputs)
        if _top:
            self.transform.wipeout_()
        return inputs
```

Why do we need to account for `__call__(self, *inputs)`?

Reply:
Again, we don't, but it makes a good UX. If I want to transform an `image` and a `bounding_box`, why do I need to put them in a container first? This is just for convenience since `image, bbox = transform(image, bbox)` is nicer than `image, bbox = transform((image, bbox))`. 

Still, let's keep the discussion narrow in this PR. Adding this would not be an issue here. If you want to discuss this further (not saying you shouldn't), please do so in #1, where I added this to the list of required features the new API should support.

---

By moving the recursive call outside of `__call__` you can avoid the `_top` parameter all together, making it more clear:

```python
class JointTransform:
    def __init__(self, transform):
        self.transform = transform
        self.transform.set_reset_auto(False)

    def _recurse(self, inputs):
        if isinstance(inputs, collections.abc.Mapping):
            return {key: self._recurse(value) for key, value in inputs.items()}
        elif isinstance(inputs, collections.abc.Sequence):
            return [self._recurse(input) for input in inputs]
        else:
            return self.transform(inputs)

    def __call__(self, inputs):
        outputs = self._recurse(inputs)
        self.transform.wipeout_()
        return outputs
```